api/konnect/v1alpha1/konnect_extension_types.go

```diff
@@ -0,0 +1,172 @@
+// extension by the DataPlane, ControlPlane or GatewayConfiguration APIs.
+// If one of the above mentioned resources successfully refers a KonnectExtension, the underlying
+// deployment(s) spec gets customized to include the konnect-related configuration.
+//
+// +genclient
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:object:generate=true
+// +kubebuilder:subresource:status
+// +kubebuilder:validation:XValidation:rule="oldSelf.spec.controlPlaneRef == self.spec.controlPlaneRef", message="spec.controlPlaneRef is immutable."
+// +kubebuilder:validation:XValidation:rule="self.spec.controlPlaneRef.type == 'konnectID' ? has(self.spec.konnect) : true",message="konnect must be set when ControlPlaneRef is set to KonnectID."
+// +kubebuilder:validation:XValidation:rule="self.spec.controlPlaneRef.type == 'konnectNamespacedRef' ? !has(self.spec.konnect) : true",message="konnect must be unset when ControlPlaneRef is set to konnectNamespacedRef."
+// +apireference:kgo:include
+// +kong:channels=gateway-operator
+type KonnectExtension struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the specification of the KonnectExtension resource.
+	Spec KonnectExtensionSpec `json:"spec,omitempty"`
+	// Status is the status of the KonnectExtension resource.
+	Status KonnectExtensionStatus `json:"status,omitempty"`
+}
+
+// KonnectExtensionList contains a list of KonnectExtension.
+//
+// +kubebuilder:object:root=true
+// +apireference:kgo:include
+type KonnectExtensionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []KonnectExtension `json:"items"`
+}
+
+// KonnectExtensionSpec defines the desired state of KonnectExtension.
+type KonnectExtensionSpec struct {
+	// ControlPlaneRef is a reference to a Konnect ControlPlane this KonnectExtension is associated with.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self.type != 'kic'", message="kic type not supported as controlPlaneRef."
+	ControlPlaneRef *commonv1alpha1.ControlPlaneRef `json:"controlPlaneRef"`
+
+	// DataPlaneClientAuth is the configuration for the client certificate authentication for the DataPlane.
+	// In case the ControlPlaneRef is of type KonnectID, it is required to set up the connection with the
+	// Konnect Platform.
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default={certificateSecret:{provisioning: Automatic}}
+	DataPlaneClientAuth *DataPlaneClientAuth `json:"dataPlaneClientAuth,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	KonnectConfiguration *KonnectConfiguration `json:"konnect,omitempty"`
+
+	// ClusterDataPlaneLabels is a set of labels that will be applied to the Konnect DataPlane.
+	//
+	// +optional
+	ClusterDataPlaneLabels map[string]string `json:"clusterDataPlaneLabels,omitempty"`
```

Comment:
Do we think it'd be a good opportunity (given that we're changing the API group) to limit this number of labels that could be set here?

Reply:
this is a good point. I've set 16 as a limit for now.

Reply:
The limit is 50 by https://github.com/Kong/aip/blob/main/src/_aips/129-labels.md#validation. Set that number.

Reply:
Good find!

I believe that given this spec we can also limit the length of keys and values (to 63) via CEL like so: https://playcel.undistro.io/?content=H4sIAAAAAAAAA3VSS0vEMBD%2BK2MP%2B4B19%2BKpeBHZg7Lqosgi1sO0ne0GJklJ0n0o%2FncnaWHZgxTCl8x8j0n6k1XEWZ4tFrAhrqwmCBbCjuB%2BuYI146lxtjP1VWGk5fIMlAc0oEwgh1VQexKN8s570iWfoLUHclQDmb1y1mgyIWrTsWXrSJh1xORUqhxU2PW%2BVmtrYHlsHXmvBK7QNB02BBPxn85jkpTmGjZOBYKT7VyKRmfOTqyHpndPSRgdIUgpYq9qgq11Er7tAtQYcCYYPu6eViDHj28vz7GuMQwq66gM49fOjNMUe%2BQO%2FzHHBpXxITmdDQah5TB%2F5FWWmeTmhGO3ooC6ZfJDMN8qh7FWyIdVJVce5owlsZ8j84RnMsY3TXgKt3ADo1G%2Fvez85K9UnhYmm2UxxkMMJC8%2B9OWFAeh7EwTAHLBHZVnmcRGmtjUJKf4sv39BWem3NAIAAA%3D%3D

This could potentially be expanded to regex match but I think the length should do for now. WDYT?

Reply:
Unresolving this one: @mlavacca WDYT about this?

Reply:
Looks much better, thanks!